noise/noise.py

```diff
@@ -0,0 +1,219 @@
+                             payment_hash=payment_hash,
+                             shared_secrets=onion['shared_secrets']
+        )
+        try:
+            plugin.rpc.waitsendpay(payment_hash=payment_hash)
+            return {'route': route, 'payment_hash': payment_hash, 'attempt': attempt}
+        except RpcError as e:
+            failcode = e.error['data']['failcode']
+            failingidx = e.error['data']['erring_index']
+            if failcode == 16399 or failingidx == len(hops):
+                return {'route': route, 'payment_hash': payment_hash, 'attempt': attempt+1}
+
+            plugin.log("Retrying delivery.")
+
+            # TODO Store the failing channel in the exclusions
+    raise ValueError('Could not reach destination {node_id}'.format(node_id=node_id))
+
+
+@plugin.async_method('sendmsg')
+def sendmsg(node_id, msg, plugin, request, pay=None, **kwargs):
+    timestamp = struct.pack("!Q", int(time.time() * 1000))
+    payload = TlvPayload()
+    payload.add_field(TLV_NOISE_MESSAGE, msg.encode('UTF-8'))
+    payload.add_field(TLV_NOISE_TIMESTAMP, timestamp)
+
+    payment_key = os.urandom(32)
+    payment_hash = hashlib.sha256(payment_key).digest()
+
+    # If we don't want to tell the recipient how to claim the funds unset the
+    # payment_key
+    if pay is not None:
+        payload.add_field(TLV_KEYSEND_PREIMAGE, payment_key)
+
+    # Signature generation always has to be last, otherwise we won't agree on
+    # the TLV payload and verification ends up with a bogus sender node_id.
+    sigmsg = hexlify(payload.to_bytes()).decode('ASCII')
+    sig = plugin.rpc.signmessage(sigmsg)
+    sigcheck = plugin.rpc.checkmessage(sigmsg, sig['zbase'])
+    sig = zbase32.decode(sig['zbase'])
+    payload.add_field(TLV_NOISE_SIGNATURE, sig)
+
+    res = deliver(
+        node_id,
+        payload.to_bytes(),
+        amt=pay if pay is not None else 10,
+        payment_hash=payment_hash
+    )
+    request.set_result(res)
+
+
+@plugin.async_method('recvmsg')
+def recvmsg(plugin, request, last_id=None, **kwargs):
+    next_id = int(last_id) + 1 if last_id is not None else len(plugin.messages)
+    if next_id < len(plugin.messages):
+        request.set_result(plugin.messages[int(last_id)].to_dict())
+    else:
+        plugin.receive_waiters.append(request)
+
+
+@plugin.hook('htlc_accepted')
+def on_htlc_accepted(onion, htlc, plugin, **kwargs):
+    payload = OnionPayload.from_hex(onion['payload'])
+    if not isinstance(payload, TlvPayload):
+        plugin.log("Payload is not a TLV payload")
+        return {'result': 'continue'}
+
+    body_field = payload.get(34349334)
+    signature_field = payload.get(34349335)
+
+    if body_field is None or signature_field is None:
+        plugin.log("Missing message body or signature, ignoring HTLC")
+        return {'result': 'continue'}
+
+    msg = Message(
+        id=len(plugin.messages),
+        sender=None,
+        body=body_field.value,
```

Comment:
Hey, as we did the mainchain test couple of days ago, I had a crash in the log in this or the next line when accessing unchecked attribute `value` on `body_field` or ` signature_field` I guess:

```
2020-01-22T21:59:38.842Z **BROKEN** lightningd: Plugin for htlc_accepted returned non-result response {"jsonrpc": "2.0", "id": 48, "error": "Error while processing htlc_accepted: AttributeError(\"'NoneType' object has no attribute 'value'\")"}
2020-01-22T21:59:38.986Z **BROKEN** lightningd: FATAL SIGNAL 6 (version v0.8.0-modded)
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: common/daemon.c:46 (send_backtrace) 0x556c67af1d69
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: common/daemon.c:54 (crashdump) 0x556c67af1db9
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x7efe067adfaf
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x7efe067adf25
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x7efe06797896
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: lightningd/log.c:819 (fatal) 0x556c67ac573c
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: lightningd/plugin_hook.c:83 (plugin_hook_callback) 0x556c67ae68e8
2020-01-22T21:59:38.986Z **BROKEN** lightningd: backtrace: lightningd/plugin.c:258 (plugin_response_handle) 0x556c67ae2b3f
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: lightningd/plugin.c:356 (plugin_read_json_one) 0x556c67ae2ce7
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: lightningd/plugin.c:388 (plugin_read_json) 0x556c67ae2e3a
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: ccan/ccan/io/io.c:59 (next_plan) 0x556c67b43445
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: ccan/ccan/io/io.c:407 (do_plan) 0x556c67b43fc6
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: ccan/ccan/io/io.c:417 (io_ready) 0x556c67b44004
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: ccan/ccan/io/poll.c:445 (io_loop) 0x556c67b461d0
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: lightningd/io_loop_with_timers.c:24 (io_loop_with_timers) 0x556c67abbfc6
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: lightningd/lightningd.c:855 (main) 0x556c67ac2d2f
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x7efe06799152
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x556c67aa73cd
2020-01-22T21:59:38.987Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0xffffffffffffffff
```

Reply:
Yep, fixed that in the last round of fixes, we now check that the necessary fields are passed in ok, otherwise we fail the HTLC.